### PR TITLE
cpu/stm32g0: add basic support + add nucleo-g070rb board

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds an initial support for the stm32g0 family of CPUs. The nucleo-g070rb board is also added so people can test.

Only a few peripheral APIs are supported for the moment: gpio (no IRQ!), timer, uart, i2c, spi, flashpage. The riotboot feature is also available.

Important features that are missing: gpio_irq, rtc

~Note that even if the clock configuration is quite similar to L4/WB families, I decided to add a new file for g0, to avoid even more ifdef there.~
The clock configuration is shared with the G4 family because they are very similar.
The clock configuration defined in the board was also adapted to allow choosing between HSE, HSI or PLL (itself clocked by either HSI or HSE) as system source clock. The PLL clock source is the default.
Since there's no HSE available on the nucleo-g070rb, the default PLL clock source is HSI (`@16MHz`). The core clock is configured to be 64MHz.

### Testing procedure

- A green Murdock
- Run `./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . nucleo-g070rb --jobs=4` and check the results are ok

The script reported these failures:
```
Failures during test:
- [tests/driver_grove_ledbar](tests/driver_grove_ledbar/test.failed)
- [tests/driver_my9221](tests/driver_my9221/test.failed)
- [tests/emcute](tests/emcute/test.failed)
- [tests/gnrc_dhcpv6_client](tests/gnrc_dhcpv6_client/test.failed)
- [tests/gnrc_dhcpv6_client_6lbr](tests/gnrc_dhcpv6_client_6lbr/test.failed)
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_ipv6_ext_opt](tests/gnrc_ipv6_ext_opt/test.failed)
- [tests/gnrc_ipv6_nib_dns](tests/gnrc_ipv6_nib_dns/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/periph_timer_short_relative_set](tests/periph_timer_short_relative_set/test.failed)
- [tests/pkg_libfixmath_unittests](tests/pkg_libfixmath_unittests/test.failed)
```

Most of them are related to network setup or root priviledge missing. `tests/pkg_libfixmath_unittests`, `tests/driver_grove_ledbar` and `tests/driver_my9221` are problematic but I haven't looked at them in detail for the moment.

### Issues/PRs references

depends on #14166

### CODE REVIEW

**Comment by aabadie** [comment]
rebased

File: cpu/stm32/cpu_common.c [hunk 1]
```diff
@@ -60,6 +64,7 @@ uint32_t periph_apb_clk(uint8_t bus)
     else {
```
Comment (benpicco) [inline cpu/stm32/cpu_common.c:64]: would

```suggestion
#ifdef CLOCK_APB2
    else {
```

also work?

**Comment by aabadie** [comment]
> do you have an idea why that is?

If I had found why, it would be fixed. I haven't looked into it recently. I'm stuck on that part. I have a branch which adapt gpio.c for g0 but it doesn't work.

> I guess it's working on G4?

Yes, G4 works exactly like other STMs for GPIO IRQs.

File: cpu/stm32/cpu_common.c (hunk 1 above)
Comment (aabadie) [reply cpu/stm32/cpu_common.c:64]: probably, but I don't know if it will be optimized the same.

Comment (benpicco) [reply cpu/stm32/cpu_common.c:64]: How about

```C
uint32_t periph_apb_clk(uint8_t bus)
{
#ifdef CLOCK_APB2
    if (bus == APB2) {
        return CLOCK_APB2;
    }
#else
    (void) bus;
#endif

    return CLOCK_APB1;
}
```

File: cpu/stm32/Makefile.features [hunk 2]
```diff
@@ -3,13 +3,16 @@ include $(RIOTCPU)/stm32/stm32_info.mk
 FEATURES_PROVIDED += bootloader_stm32
 FEATURES_PROVIDED += cpu_stm32$(CPU_FAM)
 FEATURES_PROVIDED += periph_cpuid
-FEATURES_PROVIDED += periph_gpio periph_gpio_irq
+FEATURES_PROVIDED += periph_gpio
+ifneq (g0,$(CPU_FAM))
+  FEATURES_PROVIDED += periph_gpio_irq
```
Comment (benpicco) [inline cpu/stm32/Makefile.features:8]: Should at least add a TODO here, so it's clear that this is a bug and not a hardware limitation.

Comment (aabadie) [reply cpu/stm32/Makefile.features:8]: I'll try to find the fix, maybe next week. I'd really like to have this family in but not without gpio irq.

File: cpu/stm32/cpu_common.c (hunk 1 above)
Comment (aabadie) [reply cpu/stm32/cpu_common.c:64]: It's not so simple because there's also the APB12 bus on L4/WB/G4.

Comment (benpicco) [reply cpu/stm32/cpu_common.c:64]: Yea but right now we have

```C
#if defined (CPU_FAM_STM32L4) || defined(CPU_FAM_STM32WB) || \
    defined(CPU_FAM_STM32G4)
    else if (bus == APB12) {
        return CLOCK_APB1;
    }
#endif
```

so in any case when `bus != APB2`, `CLOCK_APB1` is returned.

Comment (aabadie) [reply cpu/stm32/cpu_common.c]: You are right. I changed it that way.

**Comment by aabadie** [comment]
I found the solution for the EXTI configuration and now gpio_irq is working: tests/buttons behaves as expected

<details>

```
make BOARD=nucleo-g070rb -C tests/buttons flash term
make: Entering directory '/work/riot/RIOT/tests/buttons'
Building application "tests_buttons" for "nucleo-g070rb" with MCU "stm32".

"make" -C /work/riot/RIOT/boards/nucleo-g070rb
"make" -C /work/riot/RIOT/boards/common/nucleo
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32/bootloader
"make" -C /work/riot/RIOT/cpu/stm32/periph
"make" -C /work/riot/RIOT/cpu/stm32/stmclk
"make" -C /work/riot/RIOT/cpu/stm32/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  10112	    132	   2760	  13004	   32cc	/work/riot/RIOT/tests/buttons/bin/nucleo-g070rb/tests_buttons.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/buttons/bin/nucleo-g070rb/tests_buttons.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01293-g7c88e76a-dirty (2020-07-01-21:36)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2-1.cfg is deprecated, please switch to interface/stlink.cfg
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
srst_only separate srst_nogate srst_open_drain connect_assert_srst

Info : clock speed 2000 kHz
Info : STLINK V2J31M21 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.221340
Info : stm32g0x.cpu: hardware has 4 breakpoints, 2 watchpoints
Info : starting gdb server for stm32g0x.cpu on 0
Info : Listening on port 41327 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32g0x.cpu       hla_target little stm32g0x.cpu       reset

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x08000424 msp: 0x20000200
Info : device idcode = 0x20006460 (STM32G07/G08xx - Rev: B)
Info : flash size = 128kbytes
Info : flash mode : single-bank
Info : Padding image section 1 at 0x08002804 with 4 bytes (bank write end alignment)
Warn : Adding extra erase range, 0x08002808 .. 0x08002fff
auto erase enabled
wrote 10248 bytes from file /work/riot/RIOT/tests/buttons/bin/nucleo-g070rb/tests_buttons.elf in 0.314087s (31.863 KiB/s)

verified 10244 bytes in 0.079028s (126.587 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
shutdown command invoked
Done flashing
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2020-07-21 11:18:43,275 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
s
2020-07-21 11:18:45,957 # START
2020-07-21 11:18:45,964 # main(): This is RIOT! (Version: 2020.10-devel-284-g15b1e-pr/cpu/stm32g0_gpio_irq)
2020-07-21 11:18:45,967 # GPIO_INIT_INT pin, 1342179341
2020-07-21 11:18:45,969 # GPIO_INIT_INT pin num, 13
2020-07-21 11:18:45,971 # GPIO_INIT_INT port num, 2
2020-07-21 11:18:45,973 # On-board button test
2020-07-21 11:18:45,973 # 
2020-07-21 11:18:45,975 #  -- Available buttons: 1
2020-07-21 11:18:45,976 # 
2020-07-21 11:18:45,978 #  -- Try pressing buttons to test.
2020-07-21 11:18:45,979 # 
2020-07-21 11:18:45,980 # [SUCCESS]
2020-07-21 11:18:46,509 # Pressed BTN0
2020-07-21 11:18:46,885 # Pressed BTN0
2020-07-21 11:18:47,026 # Pressed BTN0
2020-07-21 11:18:47,230 # Pressed BTN0
2020-07-21 11:18:47,366 # Pressed BTN0
2020-07-21 11:18:48,177 # Exiting Pyterm
```

</details>

File: cpu/stm32/include/vendor/stm32g0xx.h [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  DISABLE = 0,
+  ENABLE = !DISABLE
+} FunctionalState;
+#define IS_FUNCTIONAL_STATE(STATE) (((STATE) == DISABLE) || ((STATE) == ENABLE))
+
+typedef enum
+{
+  SUCCESS = 0,
+  ERROR = !SUCCESS
+} ErrorStatus;
```
Comment (benpicco) [inline cpu/stm32/include/vendor/stm32g0xx.h:141]: I would say patch `MQTTClient.h`, but that's from an external source too :confused:

**Comment by benpicco** [comment]
Needs a rebase :smiley:

**Comment by aabadie** [comment]
Now that #14152 is merged, I adapted this one accordingly. If Murdock is green, it should be good.

**Comment by benpicco** [comment]
`examples/paho-mqtt` will still fail to compile due to conflicting global defines of `SUCCESS`

**Comment by aabadie** [comment]
> examples/paho-mqtt will still fail to compile due to conflicting global definitions of SUCCESS

It will work without any change because there's an ifdef for all STM32 (added in #14152).

**Comment by benpicco** [comment]
Please squash.

**Pull request merged**